File: packages/skillet/src/eval/schemas/benchmark.ts
import { z } from "zod";
import { ExpectationResultSchema } from "./grading.js";

const ProviderInfoSchema = z.object({
	name: z.string(),
	model: z.string(),
});

const StatsSchema = z.object({
	mean: z.number(),
	stddev: z.number(),
});

const RunResultSchema = z.object({
	pass_rate: z.number(),
	passed: z.number().int(),
	failed: z.number().int(),
	total: z.number().int(),
	time_seconds: z.number(),
	input_tokens: z.number().int(),
	output_tokens: z.number().int(),
	total_tokens: z.number().int(),
	tool_calls: z.number().int(),
	errors: z.number().int(),
	cost_usd: z.number(),
});

const BenchmarkRunSchema = z.object({
	eval_id: z.number(),
	provider: z.string(),
	model: z.string(),
	run_number: z.number().int(),
	result: RunResultSchema,
	expectations: z.array(ExpectationResultSchema),
	claims: z.array(z.string()).default([]),
	eval_feedback: z.string().nullable().default(null),
	error: z.string().nullable().default(null),
});

const ProviderSummarySchema = z.record(
	z.string(),
	z.object({
		pass_rate: StatsSchema,
		time_seconds: StatsSchema,
		total_tokens: StatsSchema,
		cost_usd: StatsSchema,
	}),
);

export const BenchmarkFileSchema = z.object({
	metadata: z.object({
		skill_name: z.string(),
		skill_path: z.string(),
		skill_version: z.string().optional(),
		skill_sha256: z.string().optional(),
		timestamp: z.string(),
		evals_run: z.array(z.number()),
		runs_per_provider: z.number().int(),
		providers: z.array(ProviderInfoSchema),
		grader: ProviderInfoSchema,
	}),
	runs: z.array(BenchmarkRunSchema),
	provider_summary: ProviderSummarySchema,
	notes: z.array(z.string()).default([]),
});

export type RunResult = z.infer<typeof RunResultSchema>;
export type BenchmarkRun = z.infer<typeof BenchmarkRunSchema>;
export type BenchmarkFile = z.infer<typeof BenchmarkFileSchema>;
export type ProviderSummary = z.infer<typeof ProviderSummarySchema>;

File: packages/skillet-eval-python/src/skill_eval/schemas/benchmark.py
from __future__ import annotations

from pydantic import BaseModel, Field

from skill_eval.schemas.grading import ExpectationResult


class ProviderInfo(BaseModel):
    name: str
    model: str


class Stats(BaseModel):
    mean: float
    stddev: float


class RunResult(BaseModel):
    pass_rate: float
    passed: int
    failed: int
    total: int
    time_seconds: float
    input_tokens: int
    output_tokens: int
    total_tokens: int
    tool_calls: int
    errors: int
    cost_usd: float


class BenchmarkRun(BaseModel):
    eval_id: int
    provider: str
    model: str
    run_number: int
    result: RunResult
    expectations: list[ExpectationResult]
    claims: list[str] = Field(default_factory=list)
    eval_feedback: str | None = None
    error: str | None = None


class BenchmarkMetadata(BaseModel):
    skill_name: str
    skill_path: str
    timestamp: str
    evals_run: list[int]
    runs_per_provider: int
    providers: list[ProviderInfo]
    grader: ProviderInfo


class ProviderSummary(BaseModel):
    pass_rate: Stats
    time_seconds: Stats
    total_tokens: Stats
    cost_usd: Stats


class BenchmarkFile(BaseModel):
    metadata: BenchmarkMetadata
    runs: list[BenchmarkRun]
    provider_summary: dict[str, ProviderSummary]
    notes: list[str] = Field(default_factory=list)
